src/data/docs/Sections/Options.rst
.. SPDX-License-Identifier: GPL-2.0-or-later

.. _Options-section:

Command Line Options
====================

As usual the help message is available via:

.. code-block::

   wg-tool --help

Which prints a list of available options.

Some items may be modified using *--edit* and some, like *--rename* have not 
edit counterpart.

Sometimes it may be more convenient to use an option rather than edit.


Current Help
------------

.. code-block:: text


    usage: wg-tool [-h] 
                   [-wkd DIR] 
                   [-migrate] 
                   [-import-configs <vpn-name>] 
                   [-edit]
                   [-copy] 
                   [-rename] 
                   [-ident vpn.account.prof]
                   [-to-ident vpn.account.prof] 
                   [-merge FILE]
                   [-nets-wanted-add NETS_WANTED_ADD]
                   [-nets-wanted-del NETS_WANTED_DEL]
                   [-nets-offered-add NETS_OFFERED_ADD]
                   [-nets-offered-del NETS_OFFERED_DEL] 
                   [-new] 
                   [-roll] 
                   [-active]
                   [-not-active] 
                   [-hidden] 
                   [-not-hidden] 
                   [-l] 
                   [-rpt RPTFILE]
                   [-rrpt] 
                   [-b] 
                   [-r] 
                   [-fp] 
                   [-v] 
                   [-V] 
                   [-hist NUM] 
                   [-hist-wg NUM]
                   [id_names ...]

    wg-tool: Manage wireguard config files

    options:
      -h, --help            show this help message and exit
      -wkd, --work-dir DIR  Set the working directory path (./)

    :*--------------------*:
    : Positional Arguments :
    :*--------------------*:
      One or more identity names ID(s)

      id_names              <vpn>.<account>.<profile> ...
                            e.g. client :	vpn1.bob.laptop
                            e.g. gateway:	vpn1.servers:gateway-1
                            NB Some options may take <vpn>.<account> or just <vpn>

    :*------------------------*:
    : Migrate & Import Options :
    :*------------------------*:
      Migrating from versions prior to 8.0.

      -migrate, --migrate   Migrate database from pre v8.0 version:
                            configs -> Data, wg-configs -> Data-wg
                            Legacy directories are unchanged
      -import-configs, --import-configs <vpn-name>
                            Import standard wireguard configs.
                             The vpn must exist with matching and networks
                             All configs should be in the work dir and
                             the path to each config file should be:
                                 <vpn-name>/<account-name>/<profile-name>.conf

    :*---------------------*:
    : Edit / Modify Options :
    :*---------------------*:
      Add, edit or modify identities.

      -edit, --edit         Edit one item.
                            Requires --ident "vpn.account.profile"
      -copy, --copy         Copy one identity to another.
                            Requires --ident <from> --to-ident <to>
      -rename, --rename     Rename an identity.
                            Requires --ident and --to-ident
      -ident, --ident vpn.account.prof
                            Specify an identity.
                            Use with --edit, --copy and --rename)
      -to-ident, --to-ident vpn.account.prof
                            Specify target identity,
                            Use with: --copy and --rename)
      -merge, --merge FILE  Merge edits from a file.
      -nets-wanted-add, --nets-wanted-add NETS_WANTED_ADD
                            Add wanted network(s) to IDs (See positional parameters)
                            Multiple networks use comma separated list
                            Special network "internet" is same as internet_wanted = true
      -nets-wanted-del, --nets-wanted-del NETS_WANTED_DEL
                            Delete wanted network(s) from IDs (See positional parameters)
                            Multiple networks as comma separated list
                            Special network "internet" is same as internet_wanted = false
      -nets-offered-add, --nets-offered-add NETS_OFFERED_ADD
                            Add offered network(s) to IDs (See positional parameters)
                            Multiple networks use comma separated list
                            Special network "internet" is same as internet_offered = true
      -nets-offered-del, --nets-offered-del NETS_OFFERED_DEL
                            Delete offered network(s) from IDs (See positional parameters)
                            Multiple networks as comma separated list
                            Special network "internet" is same as internet_offered = false
      -new, --new           Create new item (See positional parameters))]
                            Each can be one of: vpn, vpn.account or vpn.account.prof
      -roll, --roll-keys    Generate new keys for IDs (See positional parameters)
      -active, --active     Mark some IDs active. (Use positional parameters)
      -not-active, --not-active
                            Mark some IDs not active. (See positional parameters)
      -hidden, --hidden     Hidden are not shown unless double verb -vv is used.
      -not-hidden, --not-hidden
                            Mark some IDs not hidden. (See positional parameters)
      -add-ip-group, --add-ip-group Name Subnet(s)
                            Add new IP group name and subnet(s). Requires --ident <vpn-name>
                            Every network in vpn needs a group subnet given as comma separated list
                            e.g. --add-ip-group 10.77.77.28/30,fc00:77:77::28/126
      -ip-group, --ip-group Name
                            Associate specified profile as member of existing ip group <Name>
                            Requires one or more IDs on command line: <vpn>.<account>.<profile>
      -allow-ip-groups, --allow-ip-groups Name
                            Group(s) that are allowed access to this peer profile
                            Group name(s) argument is a comma separated list <group1>[,<group2>..]


    :*-----------------*:
    : Reporting Options :
    :*-----------------*:
      Peer info and reports from wg server report

      -l, --list            List vpns, accounts & profiles. May filter by providing IDs
                            profile names may comma separated list
                            e.g.
                              all : -l
                              all accounts and profiles of "vpn1": -l vpn1
                              all profiles of "vpn1.bob": -l vpn1.bob
                              2 profiles of "vpn1.bob": -l vpn1.bob.prof-1,prof-2
      -rpt, --show-rpt RPTFILE
                            Generate report from "wg show" output (file or "stdin") (See also --run-show-report)
      -rrpt, --run-show-rpt
                            Run "wg show" (must be on wireguard server) and generate report. (See also --show-rpt)
      -b, --brief           Simplified outputs. Opposite of -det

    :*---------------*:
    : General Options :
    :*---------------*:
      Miscellaneous options.

      -r, --refresh         Force a refresh and write all wireguard configs.
                            This is automatic and is not needed normally.
      -fp, --file-perms     Redo restricted file permissions on all data files. Not normally needed.
      -v, --verb            Verbose output. Repeat for even more verbosity.
                            -vv works with -l and -rpt. Use -vvv for very verbose output
      -V, --version         Version info

    :*--------------*:
    : Stored Options :
    :*--------------*:
      These are saved as new default values.

      -hist, --hist NUM     Number of previous data configs to retain (5)
      -hist-wg, --hist-wg NUM
                            Number of previous wireguard configs to retain (3)

.. _Compacting:

Compacting Networks
-------------------

Note on networks. The wireguard config *AllowedIPs* variable compacts
networks to  the minimal list of CIDRs. For convenience, the
pre-compacted list of networks is provided as a comment.

For example:

.. code-block:: none

    [10.77.77.1/32, 0.0.0.0/0, 192.168.1.0/24, ::/0, fc00:77:77::1/128]

    gets compacted to → 

    [0.0.0.0/0, ::/0]

since *192.168.1.0/24* is a subnet of *0.0.0.0/0* and similarly for the IPv6 net.

A comment is added to the config file which shows all the *AllowedIPs*
networks prior to compacting. This can be helpful when looking them over for 
testing or checking purposes.



CHANGELOG.md
# Changelog

## [5.1.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-09-25
 - README - replace markdown url links with rst link notation  
 - update CHANGELOG.md  

## [5.1.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-08-02
 - Improve code finding available client IPs to properly support IPv6.  
   Client IPs are chosen from the server Address list in natural order. If you prefer clients  
   get IPv6 addresses, those should be listed first. Similarly, if IPv4 is preferred, then put that first.  
   Tidy to keep pylint clean  
 - update CHANGELOG.md  

## [4.7.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-07-28
 - Fix import open_file buglet  
 - update CHANGELOG.md  

## [4.6.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-05-18
 - install: switch from pip to python installer package. This adds optimized bytecode  
 - update CHANGELOG.md  

## [4.5.3] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-05-18
 - PKGBUILD: build wheel back to using python -m build instead of poetry  
 - update CHANGELOG.md  

## [4.5.2] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-05-17
 - Simplify Arch PKGBUILD and more closely follow arch guidelines  
 - update CHANGELOG.md  

## [4.5.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-05-08
 - Add comment to README about linux using wg and ssh and MTU  
 - typo  
 - update CHANGELOG.md  

## [4.5.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-05-02
 - Add comment on philosophy of living at the head commit.  
   Change README from markdown to restructured text  

## [4.4.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-04-15
 - update CHANGELOG.md  
 - Only show user public key for "-rpt" when also using "-det".  
   Since we show user and profile name, the user key is not really needed  
 - update CHANGELOG.md  

## [4.3.6] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-04-11
 - postup.nft script add extra line: ct status dnat accept - I saw a martial packat at firewall from vpn which was unexpected  
   minor readme edit  
   update project version  
 - update CHANGELOG.md  

## [4.3.5] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-01-06
 - Add SPDX licensing lines  
 - update CHANGELOG.md  

## [4.3.4] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-29
 - Add reminder in README to allow ip forwarding on wireguard server  
 - update CHANGELOG.md  

## [4.3.3] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-28
 - Add brief networking note  
 - update CHANGELOG.md  

## [4.3.2] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-26
 - Change default python interpreter location to /usr/bin/python3 (remove env).  
   This is also recommended by e.g. debian packaging guidelines (https://www.debian.org/doc/packaging-manuals/python-policy). While many distros (Arch, Fedora etc.) recommend /usr/bin/python - we keep python3 which will work on those and on debian until debian provides python (and not just python3).  
 - update CHANGELOG.md  

## [4.3.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-25
 - Move archlinux dir to packaging.  
   Add packaging/requirements.txt  
   Update build dependencies in PKGBUILD  
   Tweak README  
 - tweak README  
 - update CHANGELOG.md  

## [4.3.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-20
 - Change python to python3 (as per GH issue #5 on ubuntu/debian.  
   Remove pip option from installer (--root-user-action=ignore)  
 - indent fix  
 - To help with older pre 3.9 python versions, provide files without match().  
   They are in lib38. Copy to lib38/*.py lib/  
 - update CHANGELOG.md  

## [4.2.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-14
 - update CHANGELOG.md  
 - Installer now uses pip install  
   PKGBUILD now uses poetry to build wheel  
 - update CHANGELOG.md  

## [4.1.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-08
 - Server show_rpt was not treating inactive users/profiles properly - fixed  
 - update CHANGELOG.md  

## [4.0.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-04
 - Stronger file access permissions to protect private data in configs.  
   Changes to work_dir:  
   Backward compatible with previous version.  
   Now prefers to use */etc/wireguard/wg-tool* if possible, otherwise falls back to current directory.  
   Thanks to Yann Cardon  
 - Improve comments in postup.nft including reference to alternate postup from Yann Cardon  
 - Merge: f74aa16bc2 26e957cd19  
   Merge pull request #3 from ycardon/master  
   Create postup-alternate.nft  
 - Create postup-alternate.nft  
   provides an other example of postup script with useful comments  
 - update CHANGELOG.md  

## [3.7.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-12-03
 - bug: --list if username(s) given without profile. Now we list all profiles  
 - update CHANGELOG.md  
 - Typo in README fixed by @ycardon  
 - Merge: 8c05f936df 6dcc5b6459  
   Merge pull request #2 from ycardon/master  
   small typo in the readme  
 - small typo  
   --add-users > --add_users  
 - update CHANGELOG.md  

## [3.6.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-11-30
 - bug fix for --init  
   Thanks to @ycardon - this fixes issue #1 : https://github.com/gene-git/wg_tool/issues/1  
 - update CHANGELOG.md  

## [3.5.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-11-29
 - turn off test mode  
 - update CHANGELOG.md  

## [3.4.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-11-29
 - Improve wg-peer-updn  
   - Rename existing resolv.conf when saving  
   - Add timestamp to wireguard resolv.conf  
 - update CHANGELOG.md  

## [3.3.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-11-29
 - Small add to README  
 - update CHANGELOG.md  

## [3.3.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-11-29
 - Improve README  
 - update CHANGELOG.md  

## [3.2.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-11-28
 - typo  
 - update CHANGELOG.md  


src/wg_tool/config/options.py
# SPDX-License-Identifier: GPL-2.0-or-later
# SPDX-FileCopyrightText: © 2022-present Gene C <arch@sapience.com>
# PYTHON_ARGCOMPLETE_OK
"""
Options - command line options for WgTool
"""
# pylint: disable=too-many-locals
# pylint: disable=too-many-statements
# pylint: disable=too-many-branches
from typing import (Any)
import os
import argparse

from wg_tool.utils import Msg
from wg_tool.utils import csv_string_to_list

from .options_save import (read_saved_options, write_saved_options)
from .completion import completion_init

type _Opt = tuple[str | tuple[str, ...], dict[str, Any]]


def _title_underline(title: str) -> str:
    """
    Argparse appends ":" - pretty this up a bit
    """
    n = len(title)
    lines = ':*' + n * '-' + '*'
    title = f'{lines}:\n: {title} :\n{lines}'
    return title


def _migrate_options() -> tuple[str, str, list[_Opt]]:
    """
    Migrate Group of options
    """
    opts: list[_Opt] = []

    title = _title_underline('Migrate & Import Options')
    description = 'Migrating from versions prior to 8.0.'

    txt = 'Migrate database from pre v8.0 version:\n'
    txt += 'configs -> Data, wg-configs -> Data-wg\n'
    txt += 'Legacy directories are unchanged\n'
    opts.append((('-migrate', '--migrate'),
                 {'action': 'store_true', 'help': txt}))

    txt = 'Import standard wireguard configs.\n'
    txt += ' The vpn must exist with matching and networks\n'
    txt += ' All configs should be in the work dir and\n'
    txt += ' the path to each config file should be:\n'
    txt += '     <vpn-name>/<account-name>/<profile-name>.conf\n'
    opts.append((('-import-configs', '--import-configs'),
                 {'metavar': '<vpn-name>', 'help': txt}))

    return (title, description, opts)


def _positional_options() -> tuple[str, str, list[_Opt]]:
    """
    Positional args
    """
    opts: list[_Opt] = []

    title = _title_underline('Positional Arguments')
    description = 'One or more identity names ID(s)'

    txt = ('<vpn>.<account>.<profile> ...'
           '\ne.g. client :\tvpn1.bob.laptop'
           '\ne.g. gateway:\tvpn1.servers:gateway-1'
           '\nNB Some options may take <vpn>.<account> or just <vpn>'
           )
    opts.append(('ident_names',
                {'nargs': '*', 'help': txt, 'metavar': 'id_names', }))

    return (title, description, opts)


def _edit_options() -> tuple[str, str, list[_Opt]]:
    """
    Edit Group of options
    Booleans:
        implied: modify = True (means use mods/xxx)
            --edit
            --copy
            --new
            --merge
            --rename
            --add-ip-group
            --ip-group
            --allow-ip-groups
         modify = False  (means not using mods/xxx)
            --roll-keys
            --active
            --not-active
            --hidden
            --not-hidden
     Companion options:
        --ident
        --to-idend
    """
    opts: list[_Opt] = []

    title = _title_underline('Edit / Modify Peer Options')
    description = 'Add, edit or modify identities.'

    txt = 'Edit one item.\n'
    txt += 'Requires --ident "vpn.account.profile"'
    opts.append((('-edit', '--edit'),
                 {'help': txt, 'action': 'store_true'}))

    txt = 'Copy one identity to another.\n'
    txt += 'Requires --ident <from> --to-ident <to>'
    opts.append((('-copy', '--copy'),
                 {'help': txt, 'action': 'store_true'}))

    txt = 'Rename an identity.\n'
    txt += 'Requires --ident and --to-ident'
    opts.append((('-rename', '--rename'),
                 {'help': f'{txt}', 'action': 'store_true'}))

    txt = 'Specify an identity.\n'
    txt += 'Use with --edit, --copy and --rename)'
    opts.append((('-ident', '--ident'),
                 {'help': txt, 'metavar': 'vpn.account.prof'}))

    txt = 'Specify target identity,\n'
    txt += 'Use with: --copy and --rename)'
    opts.append((('-to-ident', '--to-ident'),
                 {'help': txt, 'metavar': 'vpn.account.prof'}))

    txt = 'Merge edits from a file.'
    opts.append((('-merge', '--merge'),
                 {'metavar': 'FILE', 'help': txt}))

    # Add/Remove wanted networks
    txt = 'Add wanted network(s) to IDs (See positional parameters)\n'
    txt += 'Multiple networks use comma separated list\n'
    txt += 'Special network "internet" is same as internet_wanted = true'
    opts.append((('-nets-wanted-add', '--nets-wanted-add'),
                 {'help': txt}))

    txt = 'Delete wanted network(s) from IDs (See positional parameters)\n'
    txt += 'Multiple networks as comma separated list\n'
    txt += 'Special network "internet" is same as internet_wanted = false'
    opts.append((('-nets-wanted-del', '--nets-wanted-del'),
                 {'help': txt}))

    # Add/Remove offered networks
    txt = 'Add offered network(s) to IDs (See positional parameters)\n'
    txt += 'Multiple networks use comma separated list\n'
    txt += 'Special network "internet" is same as internet_offered = true'
    opts.append((('-nets-offered-add', '--nets-offered-add'),
                 {'help': txt}))

    txt = 'Delete offered network(s) from IDs (See positional parameters)\n'
    txt += 'Multiple networks as comma separated list\n'
    txt += 'Special network "internet" is same as internet_offered = false'
    opts.append((('-nets-offered-del', '--nets-offered-del'),
                 {'help': txt}))

    txt = 'Create new item (See positional parameters))]\n'
    txt += 'Each can be one of: vpn, vpn.account or vpn.account.prof\n'
    txt += 'A Profile can be put into ip group using --ip-group <name1>[,...]'
    opts.append((('-new', '--new'),
                 {'help': txt, 'action': 'store_true'}))

    txt = 'Generate new keys for IDs (See positional parameters)'
    opts.append((('-roll', '--roll-keys'),
                 {'action': 'store_true', 'help': txt}))

    txt = 'Mark some IDs active. (Use positional parameters)'
    opts.append((('-active', '--active'),
                 {'action': 'store_true', 'help': txt}))

    txt = 'Mark some IDs not active. (See positional parameters)'
    opts.append((('-not-active', '--not-active'),
                 {'action': 'store_true', 'help': txt}))

    txt = 'Mark some hidden. (Use positional parameters)'
    txt = 'Hidden are not shown unless double verb -vv is used.'
    opts.append((('-hidden', '--hidden'),
                 {'action': 'store_true', 'help': txt}))

    txt = 'Mark some IDs not hidden. (See positional parameters)\n'
    opts.append((('-not-hidden', '--not-hidden'),
                 {'action': 'store_true', 'help': txt}))

    #
    # Add IP Group network
    # Attach ip group to profile identity
    #   --add-ip-group create a new group
    #   --ip-group assign peer profile to this group
    #   --allow-ip-groups list of groups that are allowed access to this peer
    #
    txt = 'Add new IP group name and subnet(s). Requires --ident <vpn-name>\n'
    txt += 'Every network in vpn needs a group subnet given as comma separated list\n'
    txt += 'e.g. --add-ip-group 10.77.77.28/30,fc00:77:77::28/126'
    opts.append((('-add-ip-group', '--add-ip-group'),
                 {'nargs': 2, 'metavar': ('Name', 'Subnet(s)'),
                  'default': None, 'help': txt}))

    txt = 'Associate specified profile as member of existing ip group <Name>\n'
    txt += 'Requires one or more IDs on command line: <vpn>.<account>.<profile>'
    opts.append((('-ip-group', '--ip-group'),
                 {'metavar': 'Name', 'help': txt}))

    txt = 'Group(s) that are allowed access to this peer profile\n'
    txt += 'Group name(s) argument is a comma separated list <group1>[,<group2>..]'
    opts.append((('-allow-ip-groups', '--allow-ip-groups'),
                 {'metavar': 'Name', 'help': txt}))
    #
    # Sort options alphabetically
    #   - Sort on first as some may only have one option.
    #
    # opts.sort(key=lambda item: item[0][0])

    return (title, description, opts)


def _stored_options(saved: dict[str, Any]) -> tuple[str, str, list[_Opt]]:
    """
    Stored Group of options
    """
    opts: list[_Opt] = []

    title = _title_underline('Stored Options')
    description = 'These are saved as new default values.'

    #
    # Options that can be saved
    #
    default = '5'
    if saved and saved.get('hist'):
        default = str(saved['hist'])

    txt = f'Number of previous data configs to retain ({default})'
    opts.append((('-hist', '--hist'),
                 {'metavar': 'NUM', 'default': default, 'help': txt}))

    default = '3'
    if saved and saved.get('hist_wg'):
        default = str(saved['hist_wg'])

    txt = f'Number of previous wireguard configs to retain ({default})'
    opts.append((('-hist-wg', '--hist-wg'),
                 {'metavar': 'NUM', 'default': default, 'help': txt}))

    # toggle = True
    # if saved and saved.get('net_compact') is not None:
    #     toggle = saved['net_compact']

    # txt = 'Compact networks output in AllowedIPs'
    # opts.append((('-net-compact', '--net-compact'),
    #              {'action': 'store_true', 'default': toggle,
    #               'help': txt + f' ({toggle})'}))

    # txt = 'Do NOT ' + txt
    # opts.append((('-no-net-compact', '--no-net-compact'),
    #              {'action': 'store_false', 'dest': 'net_compact',
    #               'help': txt}))

    return (title, description, opts)


def _report_options() -> tuple[str, str, list[_Opt]]:
    """
    Reporting options.
    """
    opts: list[_Opt] = []

    title = _title_underline('Reporting Options')
    description = 'Peer info and reports from wg server report'

    txt = 'List vpns, accounts & profiles. May filter by providing IDs\n'
    txt += 'profile names may comma separated list\n'
    txt += 'e.g.\n'
    txt += '  all : -l\n'
    txt += '  all accounts and profiles of "vpn1": -l vpn1\n'
    txt += '  all profiles of "vpn1.bob": -l vpn1.bob\n'
    txt += '  2 profiles of "vpn1.bob": -l vpn1.bob.prof-1,prof-2\n'

    opts.append((('-l', '--list',),
                 {'action': 'store_true', 'help': txt}))

    txt = ('Generate report from "wg show" output (file or "stdin")'
           ' (See also --run-show-report)'
           )
    opts.append((('-rpt',  '--show-rpt'),
                 {'help': txt, 'metavar': 'RPTFILE'}))

    txt = ('Run "wg show" (must be on wireguard server) and generate report.'
           ' (See also --show-rpt)'
           )
    opts.append((('-rrpt',  '--run-show-rpt'),
                 {'action': 'store_true', 'help': txt}))

    txt = 'Simplified outputs. Opposite of -det'
    opts.append((('-b', '--brief',),
                 {'action': 'store_true', 'help': txt}))

    #
    # Sort alphabetically on first as some may only have one option.
    #
    # opts.sort(key=lambda item: item[0][0])

    return (title, description, opts)


def _general_options() -> tuple[str, str, list[_Opt]]:
    """
    Manage command line options (argparse)
    """
    opts: list[_Opt] = []

    title = _title_underline('General Options')
    description = 'Miscellaneous options.'

    txt = ('Force a refresh and write all wireguard configs.'
           '\nThis is automatic and is not needed normally.'
           )
    opts.append((('-r', '--refresh'),
                 {'action': 'store_true', 'help': txt}))

    txt = 'Redo restricted file permissions on all data files.'
    txt += ' Not normally needed.'
    opts.append((('-fp', '--file-perms'),
                 {'action': 'store_true', 'help': txt}))

    txt = 'Verbose output. Repeat for even more verbosity.\n'
    txt += '-vv works with -l and -rpt. Use -vvv for very verbose output'
    opts.append((('-v', '--verb'),
                 {'action': 'count', 'default': 0, 'help': txt}))

    # txt = 'When applicable, apply action to this vpn name.'
    # opts.append((('-vpn', '--vpn-name'),
    #              {'help': txt}))

    opts.append((('-V', '--version'),
                 {'action': 'store_true', 'help': 'Version info'}))

    #
    # Sort alphabetically on first as some may only have one option.
    #
    # opts.sort(key=lambda item: item[0][0])

    return (title, description, opts)


def parse_args(work_dir: str, data_dir: str) -> dict[str, Any]:
    """
    Command line options
     - some may be optionally saved to file in config dir
     - 2 parsing phases:
       - phase 1: parse work_dir
       - phase 2: use work_dir for remaining options.
    """
    groups: list[tuple[argparse._ArgumentGroup, list[_Opt]]] = []
    desc: str = 'wg-tool: Manage wireguard config files'

    #
    # Phase 1: "work_dir" option.
    #  -> load saved options using work_dir
    #  -> handle all remaining options.
    #
    par1 = argparse.ArgumentParser(
            add_help=False,
            formatter_class=argparse.RawTextHelpFormatter
            )

    txt = f'Set the working directory path ({work_dir})'
    par1.add_argument("-wkd", "--work-dir",
                      default=work_dir, metavar='DIR', help=txt)

    par2 = argparse.ArgumentParser(
            description=desc,
            add_help=True,
            formatter_class=argparse.RawTextHelpFormatter,
            parents=[par1])

    wkd_args, rest_args = par1.parse_known_args()
    if wkd_args.work_dir:
        work_dir = wkd_args.work_dir

    #
    # read any saved options using latest work_dir.
    #
    saved_path = os.path.join(work_dir, data_dir, 'saved_options')
    saved = read_saved_options(saved_path)

    #
    # Phase 2: Handle rest of the options
    #
    _initialize_groups(saved, par2, groups)

    #
    # Parse and return dictionary
    # Note:
    #   python 3.12 deprecated some argparse options like 'type': int
    #   - So we only use bool/str/list[str] options.
    #
    option_dict: dict[str, Any] = {}
    saved_option_dict: dict[str, Any] = {}

    completion_init(par2)
    args = par2.parse_args(rest_args, namespace=wkd_args)

    #
    # Current remaining options
    #
    if args:
        int_keys = ('hist', 'hist_wg')
        cslist_keys = ('nets_wanted_add', 'nets_wanted_del',
                       'nets_offered_add', 'nets_offered_del')
        save_keys = int_keys
        for (key, val) in vars(args).items():
            if val is not None:
                if key in int_keys:
                    option_dict[key] = int(val)

                elif key in cslist_keys:
                    option_dict[key] = csv_string_to_list(val)

                elif key == 'add_ip_group':
                    name = val[0]
                    subnet = val[1]
                    subnets = subnet.split(',')
                    option_dict['add_ip_group_name'] = name
                    option_dict['add_ip_group_subnets'] = subnets

                elif key == 'allow_ip_groups':
                    names = val.split(',')
                    option_dict[key] = names

                elif key == 'ip_group':
                    option_dict[key] = val

                else:
                    option_dict[key] = val

                if key in save_keys:
                    saved_option_dict[key] = option_dict[key]

        # tidy import-config name
        import_dir = option_dict.get('import_configs')
        if import_dir:
            if import_dir[-1] == '/':
                import_dir = import_dir[0:-1]
            if import_dir[0:2] == './':
                import_dir = import_dir[2:]
            option_dict['import_configs'] = import_dir
    #
    # Updated saved_option file if needed.
    #
    if ((not saved and saved_option_dict)
            or (saved_option_dict and saved != saved_option_dict)):
        if not write_saved_options(saved_option_dict, saved_path):
            Msg.err(f'Error saving options to {saved_path}\n')
            return option_dict
    return option_dict


def _initialize_groups(
        saved: dict[str, Any],
        par2: argparse.ArgumentParser,
        groups: list[tuple[argparse._ArgumentGroup, list[_Opt]]]):
    """
    Initialize all available option gropus
    """
    (title, desc, positional) = _positional_options()
    group_positional = par2.add_argument_group(title, desc)
    groups.append((group_positional, positional))

    (title, desc, migrate) = _migrate_options()
    group_migrate = par2.add_argument_group(title, desc)
    groups.append((group_migrate, migrate))

    (title, desc, edit) = _edit_options()
    group_edit = par2.add_argument_group(title, desc)
    groups.append((group_edit, edit))

    (title, desc, report) = _report_options()
    group_report = par2.add_argument_group(title, desc)
    groups.append((group_report, report))

    (title, desc, general) = _general_options()
    group_general = par2.add_argument_group(title, desc)
    groups.append((group_general, general))

    (title, desc, stored) = _stored_options(saved)
    group_stored = par2.add_argument_group(title, desc)
    groups.append((group_stored, stored))

    for item in groups:
        group = item[0]
        opts = item[1]

        for opt in opts:
            opt_list, kwargs = opt
            if isinstance(opt_list, str):
                group.add_argument(opt_list, **kwargs)
            else:
                group.add_argument(*opt_list, **kwargs)
